### ISSUE TITLE
Feature idea: Abort pending requests

### ISSUE THREAD

**Issue by klis87** [apollographql/apollo-client#8858]
Migrated from: apollographql/apollo-client#1541

I think that request aborts and other things from migated topics like options borrowed from Redux-Saga, for example `takeLatest` (cancelling previous request if there is new pending) could be easier implemented once `race` lands into Apollo Links, like mentioned in https://www.apollographql.com/docs/link/composition.html

**Comment by digeomel** [comment]
@joshjg I don't understand why you can't kill ALL requests using `watchQuery` + `unsubscribe`. Why can't you create a service to store all your active subscriptions and unsubscribe them there?

**Comment by joshjg** [comment]
@digeomel We have a large app using Apollo's react hooks extensively - we aren't often calling `client.query` directly where it could be swapped out easily.

**Comment by digeomel** [comment]
@joshjg apologies, it wasn't clear to me that you're using React. However, I suspect that your "ugly hack" solution can result in memory leaks, as it seems that your existing subscriptions are not properly unsubscribed.

**Comment by FreTimmerman** [comment]
> ### (Solved)
> After struggling for day i was finally able to get the solution. People who are looking for the answers to cancel the pending request, i have documented the [POC code](https://github.com/nirus/fullstack-tutorial) & solution walkthrough step by step in this [blog post](https://coder.rocks/javascript/apollo-graphql-client-abort-pending-requests) (_Read along._)
> 
> Thanks to this discussion thread. It helped a lot to arrive at the solution.
> 
> Credits to @dannycochran & @bgultekin solutions in this thread.
> 
> _Hope this helps someone out there._

This is great. but do i understand correctly that this only cancels a request when a new request with the same id gets made?
Can i cancel a request without having to send a new one?

**Comment by FreTimmerman** [comment]
@nirus 
> ### (Solved)
> After struggling for day i was finally able to get the solution. People who are looking for the answers to cancel the pending request, i have documented the [POC code](https://github.com/nirus/fullstack-tutorial) & solution walkthrough step by step in this [blog post](https://coder.rocks/javascript/apollo-graphql-client-abort-pending-requests) (_Read along._)
> 
> Thanks to this discussion thread. It helped a lot to arrive at the solution.
> 
> Credits to @dannycochran & @bgultekin solutions in this thread.
> 
> _Hope this helps someone out there._

This is great. but do i understand correctly that this only cancels a request when a new request with the same id gets made?
Can i cancel a request without having to send a new one?

**Comment by flippidippi** [comment]
Is this issue just for manually aborting requests? 

We have the issue where our variables to the `useQuery` hook are changing, but the previous request is not being canceled. Shouldn't that happen automatically?

**Comment by flippidippi** [comment]
Is this issue just for manually aborting requests?

We are having an issue where we are using `useQuery`, the variables are changing, but the previous request that is not complete is not being canceled.

**Comment by levrik** [comment]
@flippidippi We're having this exact same issue. I thought this issue is about implementing this and not about manual cancellation but you seem to be right.

Is there a tracking issue about aborting requests when variables are changing?

**Comment by flippidippi** [comment]
@levrik I'm not sure? Which is why I was asking if this issue covered this. In the original issue, it says

> cancelling previous request if there is new pending

which seems like that should be about when variables change, but most of the comments under here seem like they are addressing the more manual approach.

@hwillson should this be a separate issue?

**Comment by levrik** [comment]
@flippidippi Sorry. Question about if an issue already exists wasn't directed to you, thus the line break 😅

**Comment by dominikklein** [comment]
I there a planned in-house solution for canceling requests when the `refetch`-function is used?

**Comment by LevanArabuli** [comment]
> Hi all - as mentioned in https://github.com/apollographql/apollo-feature-requests/issues/40#issuecomment-489487888 this functionality is already supported in Apollo Client. We like the hook `cancel` suggestion in https://github.com/apollographql/apollo-feature-requests/issues/40#issuecomment-564704980 and will be looking into this. Thanks for the great discussion!

Any updates regarding this? Has this been implemented?

**Comment by dengunya** [comment]
> Hi all - as mentioned in https://github.com/apollographql/apollo-feature-requests/issues/40#issuecomment-489487888 this functionality is already supported in Apollo Client. We like the hook `cancel` suggestion in https://github.com/apollographql/apollo-feature-requests/issues/40#issuecomment-564704980 and will be looking into this. Thanks for the great discussion!

@hwillson Any update? Was it implemented? Or rejected?
It's kinda important issue without any progress for years already ;(

**Comment by maxhaice2** [comment]
Still waiting for updates @hwillson
